### PR TITLE
xSQLServerSetup: Bug fixes that prevented Get-DscConfiguration to execute properly

### LINKED ISSUE #393: xSQLServerSetup: Get-DscConfiguration fails with error 'GetConfiguration did not succeed.'
**Details of the scenario you try and problem that is occurring:**
Get-DscConfiguration fails with error 'GetConfiguration did not succeed.'

```
PS C:\WINDOWS\system32> Get-DscConfiguration -CimSession $session -Verbose
VERBOSE: sqltest.company.local: An LCM method call arrived from computer SQLTEST with user sid S-1-5-21-2246212248-426023572-517575333-1441.
VERBOSE: sqltest.company.local: [SQLTEST]: LCM:  [ Start  Get      ]
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[WindowsFeature]NetFramework35] The operation 'Get-WindowsFeature' started: NET-Framework-Core
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[WindowsFeature]NetFramework35] The operation 'Get-WindowsFeature' succeeded: NET-Framework-Core
VERBOSE: sqltest.company.local: [SQLTEST]: LCM:  [ End    Get      ]  [[WindowsFeature]NetFramework35]  in 0.2500 seconds.
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[WindowsFeature]NetFramework45] The operation 'Get-WindowsFeature' started: NET-Framework-45-Core
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[WindowsFeature]NetFramework45] The operation 'Get-WindowsFeature' succeeded: NET-Framework-45-Core
VERBOSE: sqltest.company.local: [SQLTEST]: LCM:  [ End    Get      ]  [[WindowsFeature]NetFramework45]  in 0.1880 seconds.
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xPowerPlan]SetPlanHighPerformance] Perform operation 'Query CimInstances' with following parameters, ''queryExpression' = SELECT * FROM Win32_PowerPlan WHERE ElementName = 'Balanced','queryDi
alect' = WQL,'namespaceName' = root\cimv2\power'.
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xPowerPlan]SetPlanHighPerformance] Operation 'Query CimInstances' complete.
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xPowerPlan]SetPlanHighPerformance] The power plan 'Balanced' is the active plan.
VERBOSE: sqltest.company.local: [SQLTEST]: LCM:  [ End    Get      ]  [[xPowerPlan]SetPlanHighPerformance]  in 0.0150 seconds.
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] 2017-02-18_14-22-35: Using path: C:\Windows\TEMP\SQL2014SP1-ENT\setup.exe
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] Perform operation 'Query CimInstances' with following parameters, ''queryExpression' = SELECT * FROM Win32_Service WHERE Name = 'MSSQLSERVER','queryDia
lect' = WQL,'namespaceName' = root\cimv2'.
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] Operation 'Query CimInstances' complete.
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] Perform operation 'Query CimInstances' with following parameters, ''queryExpression' = SELECT * FROM Win32_Service WHERE Name = 'SQLSERVERAGENT','query
Dialect' = WQL,'namespaceName' = root\cimv2'.
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] Operation 'Query CimInstances' complete.
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] 2017-02-18_14-22-35: Detecting replication feature (HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\MSSQL12.MSSQLSERVER\ConfigurationState)
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] 2017-02-18_14-22-35: Replication feature not detected
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] 2017-02-18_14-22-35: Client Connectivity Tools feature not detected
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] 2017-02-18_14-22-35: Client Connectivity Tools Backwards Compatibility feature not detected
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] 2017-02-18_14-22-35: Connected to SQL localhost
VERBOSE: sqltest.company.local: [SQLTEST]:                            [[xSQLServerSetup]InstallDefaultInstance] 2017-02-18_14-22-35: Clustered instance not detected
Get-DscConfiguration : GetConfiguration did not succeed.
At line:1 char:1
+ Get-DscConfiguration -CimSession $session -Verbose
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (MSFT_DSCLocalConfigurationManager:root/Microsoft/...gurationManager) [Get-DscConfiguration], CimException
    + FullyQualifiedErrorId : MI RESULT 1,Get-DscConfiguration
    + PSComputerName        : sqltest.company.local
```

**The DSC configuration that is using the resource (as detailed as possible):**
```
xSQLServerSetup $resourceConfigName
{
    InstanceName = $Node."$($currentSqlInstance)InstanceName"
    Features = $Node."$($currentSqlInstance)InstallFeatures"
    SQLCollation = 'SQL_Latin1_General_CP1_CI_AS'
    SQLSvcAccount = $SqlServiceCredential
    AgtSvcAccount = $SqlAgentServiceCredential
    ASSvcAccount = $SqlServiceCredential
    SQLSysAdminAccounts = 'COMPANY\SQL Administrators', $SqlAdministratorCredential.UserName
    SetupCredential = $SqlInstallCredential

    InstallSharedDir = "C:\Program Files\Microsoft SQL Server"
    InstallSharedWOWDir = "C:\Program Files (x86)\Microsoft SQL Server"
    InstanceDir = "C:\Program Files\Microsoft SQL Server"
    InstallSQLDataDir = "C:\Program Files\Microsoft SQL Server\MSSQL11.MSSQLSERVER\MSSQL\Data"
    SQLUserDBDir = "C:\Program Files\Microsoft SQL Server\MSSQL11.MSSQLSERVER\MSSQL\Data\User"
    SQLUserDBLogDir = "C:\Program Files\Microsoft SQL Server\MSSQL11.MSSQLSERVER\MSSQL\Data\User"
    SQLTempDBDir = "C:\Program Files\Microsoft SQL Server\MSSQL11.MSSQLSERVER\MSSQL\Data\Temp"
    SQLTempDBLogDir = "C:\Program Files\Microsoft SQL Server\MSSQL11.MSSQLSERVER\MSSQL\Data\Temp"
    SQLBackupDir = "C:\Program Files\Microsoft SQL Server\MSSQL11.MSSQLSERVER\MSSQL\Data\Backup"

    SourcePath = $Node."SourcePath$($currentSqlInstance)"

    UpdateEnabled = 'False'

    ForceReboot = $true

    DependsOn = '[WindowsFeature]NetFramework35','[WindowsFeature]NetFramework45'
}
```
**Version of the Operating System, SQL Server and PowerShell the DSC Target Node is running:**
Windows Server 2012 R2, SQL Server 2014 SP1, PowerShell 5.0

**Version of the DSC module you're using, or 'dev' if you're using current dev branch:**
Dev

### LINKED ISSUE #661: xSQLServerScript: Get-DscConfiguration fails
**Details of the scenario you tried and the problem that is occurring:**
Deployed SQL with xSQLServerSetup and configured it with xSQLServerScript. Running a Get-DscConfiguration returned an error when it got to the xSqlServerScript section.
**The DSC configuration that is using the resource (as detailed as possible):**

**Version of the Operating System, SQL Server and PowerShell the DSC Target Node is running:**
Server 2012 R2, SQL Server 2012, PS v5.0
**What module (SqlServer or SQLPS) and which version of the module the DSC Target Node is running:**
SqlServer
**Version of the DSC module you're using, or 'dev' if you're using current dev branch:**
7.1.0.0

### PR DESCRIPTION
**Pull request by aromano2**
**Pull Request (PR) description**
Bug fixes that prevented Get-DscConfiguration to execute properly

**This Pull Request (PR) fixes the following issues:**
Fixes #393 
Fixes #661 
**Task list:**

- [ ] Change details added to Unreleased section of CHANGELOG.md?
- [ ] Added/updated documentation, comment-based help and descriptions in .schema.mof files where appropriate?
- [ ] Examples appropriately updated?
- [ ] New/changed code adheres to [Style Guidelines](https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md)?
- [ ] [Unit and (optional) Integration tests](https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md) created/updated where possible?

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/xsqlserver/658)

### CODE REVIEW

**Comment by johlju** [comment]
@aromano2 Thanks for sending in this PR! I will review now.

**Comment by johlju** [comment]
Could you please update the Unreleased section of the CHANGELOG.md with the changes.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[DSCResources/MSFT_xSQLServerScript/MSFT_xSQLServerScript.psm1, line 77 at r1](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-KnPzeWE2OqX1rn1w5qX:-KnPzeWE2OqX1rn1w5qY:b-m4k242) ([raw file](https://github.com/powershell/xsqlserver/blob/e930255246fbba2745659c2cbc901e822f5af55f/DSCResources/MSFT_xSQLServerScript/MSFT_xSQLServerScript.psm1#L77)):*
> ```PowerShell
>         GetFilePath = [System.String] $GetFilePath
>         TestFilePath = [System.String] $TestFilePath
>         Credential = [System.Object] $Credential
> ```

Could you please submit a an issue for this change? And also add that issue to the list of fixed issues in the PR description.

---

*[DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1, line 118 at r1](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-KnPz1ba_FLIrCgdoMee:-KnPz1bbVzgRqVGTVCOu:b-dnabgs) ([raw file](https://github.com/powershell/xsqlserver/blob/6514c9b53a621260dc78902060e9f4ef73481529/DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1#L118)):*
> ```PowerShell
> 
>     $features = ''
>     $clusteredSqlGroupName = ''
> ```

Curious. When these three variables were set to blank how did that impact issue #393?
I would think Get-DscConfiguration would return an empty string? Or is that the problem that a string is not allowed to be empty when it is returned, but $null is accepted?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-:-KnQ-Vk1nNgY5EpNLUTr:bed7atp)*

**Comment by aromano2** [comment]
Updated as requested

**Comment by johlju** [comment]
Could you please go into Reviewable and write or click Done on each review comment? Then I can acknowledge (resolve) the comments. Thanks. 

You find Reviewable on the big Purple button in the PR description.

**Comment by aromano2** [comment]
:lgtm:

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[DSCResources/MSFT_xSQLServerScript/MSFT_xSQLServerScript.psm1, line 77 at r1](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-KnPzeWE2OqX1rn1w5qX:-KnQmTBeb6RgdBAdQALr:b-896fix) ([raw file](https://github.com/powershell/xsqlserver/blob/e930255246fbba2745659c2cbc901e822f5af55f/DSCResources/MSFT_xSQLServerScript/MSFT_xSQLServerScript.psm1#L77)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Could you please submit a an issue for this change? And also add that issue to the list of fixed issues in the PR description.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1, line 118 at r1](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-KnPz1ba_FLIrCgdoMee:-KnQmQGVPeHerAdey4_K:b-896fix) ([raw file](https://github.com/powershell/xsqlserver/blob/6514c9b53a621260dc78902060e9f4ef73481529/DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1#L118)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Curious. When these three variables were set to blank how did that impact issue #393?
I would think Get-DscConfiguration would return an empty string? Or is that the problem that a string is not allowed to be empty when it is returned, but $null is accepted?

</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-:-KnQmX-UuqY8OTaIEEod:bnfp4nl)*

**Comment by johlju** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-:-KnV-tK79Gb3uo19h1A9:bnfp4nl)*

**Comment by johlju** [comment]
I'm in London his week so will merge this one as soon as I have the opportunity.

**Comment by johlju** [comment]
@aromano2 Just a few more review comments. Sorry for this, but only had my phone to review in London, now when I got back to my screens I saw that the change log could be improved. :smile: 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[CHANGELOG.md, line 25 at r2](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-KnrV-taqn-Bh707WFIX:-KnrV-taqn-Bh707WFIY:b54tbxi) ([raw file](https://github.com/powershell/xsqlserver/blob/2c809daf833655a2d84fb8d96aab4ce643da7c08/CHANGELOG.md#L25)):*
> ```Markdown
>   - Updated Get-TargetResource to correctly detect BOL, Conn, BC and other tools when they are installed without SQLENGINE (issue #591).
>   - Now it can detect Documentation Components correctly after the change in issue #591 (issue #628)
>   - Fixed bug that would prevent a successful return of Get-DscConfiguration if a cluster install is not used. (issue #393)
> ```

I see you got more detailed explanation in the commit message. Could you please add them here as well (and maybe rephrase accordingly).

Minor, add '.' to the end of the sentence (moving '.' after '(issue #393)'.

---

*[CHANGELOG.md, line 27 at r2](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-KnrVLURuMg-ceUnNyx7:-KnrVLURuMg-ceUnNyx8:b-8ee0yd) ([raw file](https://github.com/powershell/xsqlserver/blob/2c809daf833655a2d84fb8d96aab4ce643da7c08/CHANGELOG.md#L27)):*
> ```Markdown
>   - Fixed bug that would prevent a successful return of Get-DscConfiguration if a cluster install is not used. (issue #393)
> - Changes to xSQLServerScript
>   - Fixed bug with schema and variable mismatch (issue #661)
> ```

I see you got more detailed explanation in the commit message. Could you please add them here as well (and maybe rephrase accordingly).

Please add a full stop '.' after the sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-:-KnrVZwYQSXi9wUrjZGk:b7sge0w)*

**Comment by johlju** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/658#-:-KntrgqFuM8qhhKaJURj:bnfp4nl)*
